gateway-sp-comms/src/single_sp/update.rs

```diff
@@ -662,11 +664,112 @@ async fn send_update_in_chunks(
             "offset" => offset,
         );
 
-        image = send_single_update_chunk(cmds_tx, component, id, offset, image)
-            .await?;
+        let result;
+        (image, result) =
+            send_single_update_chunk(cmds_tx, component, id, offset, image)
+                .await;
 
-        // Update our offset according to how far our cursor advanced.
-        offset += (image.position() - prior_pos) as u32;
+        match result {
+            Ok(()) => {
+                // Update our offset according to how far our cursor advanced.
+                offset += (image.position() - prior_pos) as u32;
+            }
+            Err(
+                err @ CommunicationError::SpError(SpError::InvalidUpdateChunk),
+            ) => {
+                warn!(
```

Comment:
The logic here seems good, although we're creeping right-ward – maybe move this into a separate function?  That function could also return its own error type, so we could consolidate
```
error!(
    log,
    "invalid update chunk recovery failed: {inner_err}"
);
```
into one place

Reply:
oh it looks like the logging is structured, so maybe that consolidation doesn't make sense (but moving to a function still seems reasonable)

Reply:
Agreed; moved out in [e813215](https://github.com/oxidecomputer/management-gateway-service/pull/334/commits/e813215a7b0234e5c34c31eb7145727a8dbcc663)